### PR TITLE
Closes #217: Add microbiology related SDTM (MB, MS, BE)

### LINKED ISSUE #217: Feature Request:  Add Microbiology domains (MB, MS)
### Feature Idea

Would be nice to incorporate new linked SDTM: MB (Microbiology) and MS (Susceptibility) test data. Optionally could also be interested to link it with BE (Biospecimen Events).

# List of ToDo's
Properly linked and synthetic test data for:
- [ ] MB (Microbiology) test data
- [ ] MS (Susceptibility) test data
- [ ] BE (Biospecimen Events) test data

### Relevant Output
For dataset details check CDISC standards: SDTMIG v3.4-FINAL_2022_07_21

### PR DESCRIPTION
**Pull request by Gero1999**
> Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   
Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

## Summary of the implementation

The PR introduces new synthetic microbiology datasets to the package, specifically adding the Biospecimen Events (BE), Microbiology Findings (MB), and Microbiology Susceptibility (MS) SDTM domains.

All the microbiology data utilized is in `mb.R` as a nested list (`study_microb_data`). The nested list structure follows the material process on how the data was collected in the lab (`patient` > `visit for sample collection` > `aliquoting` > `culture` > `MB & MS tests`). In order to standardize and simplify the generation of test results for MB and MS, helper functions have been created and can be used to generate new MB/MS test results inside `study_microb_data`.

In `mb.R` a nested loop will read `study_microb_data` and derive from it the corresponding SDTM variables for each domain (MB, MS, BE). This ensures that all variables across domains are correctly linked. This file is then sourced in `ms.R` & `be.R` to get the microbiology data. Each file is also responsible of ordering and labelling its own domain variables (`MB - mb.R`, `MS - ms.R`,  `BE - be.R`), as well as of saving the object in `data/`. 

## Checklist
- [X] Place Closes #217  into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by Fanny-Gautier** [comment]
Following tick boxes to be ticked as Not Applicable for `{pharmaversesdtm}`
- Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
 - If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
 - Address any updates needed for vignettes and/or templates

Following tick boxes to apply and tick once completed without any error/warning:
- Build admiral site pkgdown::build_site() and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page.
We should read here **pharmaversesdtm** site, but still `pkgdown::build_site()` to be run.
- Link the issue in the Development Section on the right hand side.
- Pat yourself on the back for a job well done! Much love to your accomplishment!

File: data-raw/mb.R [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+    MBLNKGRP = "Link Group ID",
+    MBMETHOD = "Method of Test or Examination",
+    MBREFID = "Reference ID",
+    MBTESTCD = "Microbiology Test or Finding Short Name",
+    MBTEST = "Microbiology Test or Finding Name",
+    MBTSTDTL = "Measurement, Test or Examination Detail",
+    MBORRES = "Original Result",
+    MBORRESU = "Original Units",
+    MBRSLSCL = "Result Scale",
+    MBSTRESC = "Character Result/Finding in Standard Format",
```
Comment (Fanny-Gautier) [inline data-raw/mb.R]: Ensure to label `MBSTRESC` as `Character Result/Finding in Std Format`

File: R/mb.R [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#'     \item{MBORRES}{Original Result}
+#'     \item{MBORRESU}{Original Units}
+#'     \item{MBRSLSCL}{Result Scale}
+#'     \item{MBSTRESC}{Character Result/Finding in Standard Format}
+#'     \item{MBSTRESN}{undocumented field}
+#'     \item{MBSTRESU}{undocumented field}
+#'     \item{MBSPEC}{Specimen Type}
+#'     \item{MBLOC}{Specimen Collection Location}
+#'     \item{MBMETHOD}{Method of Test or Examination}
+#'     \item{VISITNUM}{Visit Number}
```
Comment (Fanny-Gautier) [inline R/mb.R]: `VISIT` variable seems missing. Is it on purpose?

File: data-raw/mb.R [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+            )
+          )
+        )
+      )
+    )
+  )
+)
+
+
+# ============================================================
```
Comment (Fanny-Gautier) [inline data-raw/mb.R]: Please refer to the [programming strategy - Comment section](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#source-code-formatting) to structure the file.

File: data-raw/mb.R [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+            )
+          ms <- dplyr::bind_rows(ms, res)
+        }
+      }
+    }
+  }
+}
+
+
+############################################################
```
Comment (Fanny-Gautier) [inline data-raw/mb.R]: Please refer to the [programming strategy - Comment section](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#source-code-formatting) to structure the file.

File: data-raw/be.R [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# This file only processes and saves the BE domain from that shared data source.
+# Dataset: BE
+# A synthetic SDTM BE domain representing specimen collection, aliquoting
+# and culturing events with linkage to MB (mb) and MS (ms) domains
```
Comment (Fanny-Gautier) [inline data-raw/be.R]:
```suggestion
# and culturing events with linkage to Microbiology Specimen (mb) and Microbiology Susceptibility (ms) domains
```

File: data-raw/ms.R [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    MSREFID = "Reference ID",
+    NHOID = "Name of Organism Identified",
+    MSGRPID = "Group ID",
+    MSLNKID = "Link ID",
+    MSTESTCD = "Microbial Susceptibility Test Short Name",
+    MSTEST = "Microbial Susceptibility Test Name",
+    MSAGENT = "Agent Name",
+    MSCONC = "Agent Concentration",
+    MSCONCU = "Agent Concentration Units",
+    MSORRES = "Original Result",
```
Comment (Fanny-Gautier) [inline data-raw/ms.R]:
```suggestion
    MSORRES = "Result or Finding in Original Units",
```

File: data-raw/ms.R [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    MSORRESU = "Original Units",
+    MSSTRESC = "Result or Finding in Standard Format",
+    MSSTRESN = "Numeric Result/Finding in Standard Units",
+    MSSTRESU = "Standard Units",
+    MSSPEC = "Specimen Type",
```
Comment (Fanny-Gautier) [inline data-raw/ms.R]:
```suggestion
    MSSPEC = "Specimen Material Type",
```

File: data-raw/ms.R [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    MSLOC = "Specimen Collection Location",
```
Comment (Fanny-Gautier) [inline data-raw/ms.R]:
```suggestion
    MSLOC = "Location Used for the Measurement",
```

File: data-raw/ms.R [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    MSMETHOD = "Method of Test or Examination",
+    VISITNUM = "Visit Number",
+    MSDTC = "Date/Time of Specimen Collection"
```
Comment (Fanny-Gautier) [inline data-raw/ms.R]:
```suggestion
    MSDTC = "Date/Time of Collection"
```

File: .github/CODEOWNERS [hunk 10]
```diff
@@ -35,3 +35,8 @@ data-raw/nv_neuro.R @jwang-lilly
 data-raw/suppnv_neuro.R @jwang-lilly
 data-raw/ag_neuro.R @jwang-lilly
 data-raw/lb_neuro.R @jackychen651
+
+# others
+data-raw/be.R @Gero1999
+data-raw/mb.R @Gero1999
+data-raw/ms.R @Gero1999
```
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply .github/CODEOWNERS]: This should now be fixed since I added you to the repo with "write" access.

File: data-raw/mb.R (hunk 4 above)
Comment (Gero1999) [reply data-raw/mb.R]: I made a commit for this, lmk if that would be sufficient

File: data-raw/mb.R [hunk 11]
```diff
@@ -0,0 +1,846 @@
...
+        )
+      )
+    )
+  )
+)
+
+
+# Extraction Loop: Build BE, MB, MS Domains ----
+
+studyid <- "MI"
```
Comment (Fanny-Gautier) [inline data-raw/mb.R]: For `STUDYID` and `USUBJID` variables, please use the existing ones from `pharmaversesdtm::dm`.

File: data-raw/mb.R [hunk 12]
```diff
@@ -0,0 +1,848 @@
...
+        )
+      )
+    )
+  )
+)
+
+
+# Extraction Loop: Build BE, MB, MS Domains ----
+dm <- pharmaversesdtm::dm
+studyid <- unique(dm$STUDYID)[1]
```
Comment (Gero1999) [inline data-raw/mb.R]: hey @Fanny-Gautier regarding this I hope is ok if I go back to a made-up name (e.g, "XYZ", similar to the nomenclature used in `dm_vaccine`). I am just a bit concerned that the name can confuse someone and make them think that this dataset comes from the `CDISCPILOT01`

**Comment by Lina2689** [comment]
@Gero1999, Please finish this PR so that it can be merged before the planned release at the end of March.
